include/seqan3/io/alignment_file/input.hpp

```diff
@@ -0,0 +1,1142 @@
+    detail::TypeListOfAlignmentFileInputFormats  valid_formats_ = type_list<alignment_file_format_sam>,
+    char_concept                                 stream_char_type_ = char>
+class alignment_file_input
+{
+public:
+    /*!\name Template arguments
+     * \brief Exposed as member types for public access.
+     * \{
+     */
+    //!\brief A traits type that defines aliases and template for storage of the fields.
+    using traits_type           = traits_type_;
+    //!\brief A seqan3::fields list with the fields selected for the record.
+    using selected_field_ids    = selected_field_ids_;
+    //!\brief A seqan3::type_list with the possible formats.
+    using valid_formats         = valid_formats_;
+    //!\brief Character type of the stream(s), usually `char`.
+    using stream_char_type      = stream_char_type_;
+    //!\}
+
+private:
+    //!\brief The dummy ref sequence type if no reference information were given.
+    using dummy_ref_type = decltype(ranges::view::repeat_n(typename traits_type::sequence_alphabet{}, size_t{}) |
+                                    std::view::transform(detail::access_restrictor_fn{}));
+public:
+    /*!\name Field types and record type
+     * \brief These types are relevant for record/row-based reading; they may be manipulated via the \ref traits_type
+     * to achieve different storage behaviour.
+     * \{
+     */
+    //!\brief The type of field::SEQ (default std::vector<seqan3::dna5>).
+    using sequence_type            = typename traits_type::template sequence_container<
+                                         typename traits_type::sequence_alphabet>;
+    //!\brief The type of field::ID (default std::string by default).
+    using id_type                  = typename traits_type::template id_container<
+                                         typename traits_type::id_alphabet>;
+    //!\brief The type of field::OFFSET is fixed to int32_t.
+    using offset_type              = int32_t;
+    /*!\brief The type of field::REF_SEQ (default depends on construction).
+     *
+     * If no reference information are given on construction, this type deduces to a sized view that throws on
+     * access (since there is nothing to access anyway). If the reference information are given, the type is deduced
+     * to a view over the given input reference sequence type such that no sequence information is copied.
+     */
+    using ref_sequence_type = std::conditional_t<std::Same<typename traits_type::ref_sequences, ref_info_not_given>,
+                                                 dummy_ref_type,
+                                                 decltype(std::declval<
+                                                    detail::transformation_trait_or_t<
+                                                          seqan3::reference<typename traits_type::ref_sequences const>,
+                                                        dummy_ref_type> /* does not matter as type is not chosen */
```

Comment:
why do you need this?
Does it need to be a valid expression for both cases (ref info given / not given)?

Reply:
yes. If ref info is not given, the type resolves the dummy seq type which is a repeat_n | transform. If ref info is given I deduce the type from the parameter into the traits type.